gemrb/core/GlobalTimer.cpp

```diff
@@ -249,58 +250,32 @@ void GlobalTimer::SetFadeFromColor(tick_t Count, unsigned short factor)
 
 void GlobalTimer::AddAnimation(SpriteAnimation* anim)
 {
-	// if there are no free animation reference objects,
-	// alloc one, else take the first free one
-	if (first_animation != 0) {
-		animations.erase (animations.begin());
-		first_animation--;
-	}
-
-	// and insert it into list of other anim refs, sorted by time
-	auto it = animations.begin() + first_animation;
-	for (; it != animations.end (); ++it) {
-		if ((*it)->Time() > anim->Time()) {
-			animations.insert(it, anim);
-			break;
-		}
-	}
-	if (it == animations.end())
-		animations.push_back(anim);
+	animations.insert(anim);
 }
 
 void GlobalTimer::RemoveAnimation(SpriteAnimation* anim)
 {
-	// Animation refs for given control are not physically removed,
-	// but just marked by erasing ptr to the control. They will be
-	// collected when they get to the front of the vector
-	for (auto it = animations.begin() + first_animation; it != animations.end (); ++it) {
+	for (auto it = animations.lower_bound(anim); it != animations.end() && (*it)->Time() == anim->Time(); ++it) {
```

Comment:
I'm not sure we should be checking the time here. shouldnt we remove the animation regardless of the time?

Reply:
Looking up by time first brings us to the matching slice of the set in logarithmic time. Otherwise, we to potentially iterate through the whole container to find the pointer -- under the assumption that there will never be a discrepancy in the set's tree and the actual time in the struct.

Reply:
sure, you look up using `lower_bound` thats fine, I was talking about the `(*it)->Time() == anim->Time();` part. Is it valid to have the same animation in the collection multiple times? that seems like it would cause mysterious bugs with animations being advanced multiple times a frame, or out of sync with their internal timing.

Reply:
I just wanted to make sure not to exceed the loop if `lower_bound` brings us above the timestamp we are looking for, because something isn't just in there.

But no, the same animation in terms of its location is not supposed to happe.

Reply:
that makes sense